### PR TITLE
[OP-66] - Add extra admin, accounts and nodes at contract deployment 

### PR DESCRIPTION
**Pull request by usmansaleem**
-- Added 0_ migration script to perform validation for extra admin, account and nodes during initial contract initialisation. 
 -- Modified other migration script to add admin, whitelisted nodes and accounts respectively during contract initialisation.
-- Added new module that is used by 0_ migration script.
-- Moved Admin and Account events to super class to allow addAll to emit events (contracts). 
-- Also added addAll method for AccountList (contracts).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -63,8 +63,13 @@ The deployment process covers 3 steps:
   - `PANTHEON_NODE_PERM_ACCOUNT`: The address of the account that will be used to deploy the contracts
   - `PANTHEON_NODE_PERM_KEY`: The private key associated with the deploying account's address
   - `PANTHEON_NODE_PERM_ENDPOINT`: The json rpc url endpoint that can be used to communicate with your Pantheon node
-2. If this is the first time setting up the project, run `yarn install` to initialise project dependencies, otherwise skip this step. 
-3. With these environment variables provided run `truffle migrate --reset` to deploy the contracts
+
+2. Following environment variables are optional and can be used to whitelist account and nodes during initial contract deployment 
+  - `INITIAL_ADMIN_ACCOUNTS`: The admin account addresses. Comma separated multiple addresses can be specified
+  - `INITIAL_WHITELISTED_ACCOUNTS`: The whitelisted account addresses. Comma separated multiple addresses can be specified
+  - `INITIAL_WHITELISTED_NODES`: The enode URL of whitelisted nodes. Comma separated multiple nodes can be specified
```
Comment (macfarla) [inline README.md]:
```suggestion
  - `INITIAL_WHITELISTED_NODES`: The enode URLs of whitelisted nodes. Comma separated multiple nodes can be specified
```

File: migrations/3_deploy_node_ingress_rules_contract.js [hunk 2]
```diff
@@ -34,6 +71,47 @@ module.exports = async(deployer, network) => {
...
+    //add additional whitelisted nodes
+    if (initialWhitelistedNodes) {
+        let initialWhitelistedNodesList = initialWhitelistedNodes.split(/,/).map(n => n.trim());
+
+        //Convert to enode structure
+        if(initialWhitelistedNodesList && initialWhitelistedNodesList.length > 0) {
+            for (i=0; i < initialWhitelistedNodesList.length; i++) {
+                let enode = initialWhitelistedNodesList[i];
+                if (!isValidEnode(enode)) {
+                    console.log("   > NodeRules Contract - Bypassing invalid ENode URL: " + enode);
```
Comment (macfarla) [inline migrations/3_deploy_node_ingress_rules_contract.js]:
```suggestion
                    console.log("   > NodeRules Contract - Skipping invalid eNode URL: " + enode);
```

File: README.md [hunk 3]
```diff
@@ -63,8 +63,13 @@ The deployment process covers 3 steps:
   - `PANTHEON_NODE_PERM_ACCOUNT`: The address of the account that will be used to deploy the contracts
   - `PANTHEON_NODE_PERM_KEY`: The private key associated with the deploying account's address
   - `PANTHEON_NODE_PERM_ENDPOINT`: The json rpc url endpoint that can be used to communicate with your Pantheon node
-2. If this is the first time setting up the project, run `yarn install` to initialise project dependencies, otherwise skip this step. 
-3. With these environment variables provided run `truffle migrate --reset` to deploy the contracts
+
+2. Following environment variables are optional and can be used to whitelist account and nodes during initial contract deployment
```
Comment (macfarla) [inline README.md]:
```suggestion
2. The following environment variables are optional and can be used to whitelist accounts and nodes during initial contract deployment. 
```

File: contracts/AccountRulesList.sol [hunk 4]
```diff
@@ -76,4 +101,36 @@ contract AccountRulesList {
...
+                    allAddresses[counter++] = address(pointer);
+                }
+
+                if (next != 0) {
+                    pointer = next;
+                } else {
+                    hasNext = false;
+                }
+            }
+            //Getting rid of unused variable warning
```
Comment (macfarla) [inline contracts/AccountRulesList.sol]: did we not have a better way of doing this?

Comment (usmansaleem) [reply contracts/AccountRulesList.sol]: Copied from AdminList.sol ...

**Comment by usmansaleem** [comment]
Will re-open after the changes

File: contracts/AccountRulesList.sol [hunk 5]
```diff
@@ -32,6 +41,22 @@ contract AccountRulesList {
         }
     }
 
+    function addAll(address[] memory accounts) internal returns (bool) {
+        bool allAdded = true;
+        for (uint i = 0; i<accounts.length; i++) {
+            if (!exists(accounts[i])) {
```
Comment (Errorific) [inline contracts/AccountRulesList.sol]: Don't think you need the if statement here, I think you can change this to be something like 
```
bool added = add(accounts[i]);
emit AccountAdded(added, accounts[i]);
allAdded = allAdded && added;
```

Comment (usmansaleem) [reply contracts/AccountRulesList.sol]: 👍

File: scripts/whitelist_utils.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
+const Web3Utils = require("web3-utils");
+
+/* Optional initial Whitelisted Nodes */
+let envInitialWhitelistedNodes = process.env.INITIAL_WHITELISTED_NODES;
```
Comment (macfarla) [inline scripts/whitelist_utils.js]: for consistency should all 3 be initialised here?

Comment (usmansaleem) [reply scripts/whitelist_utils.js]: @macfarla thx for picking it, I was meant to move them ... they are now moved inside respective methods.

**Pull request merged**